### PR TITLE
Feature skip double rtrx 15

### LINKED ISSUE #15: skip Triplicate ACK when ack is ahead of retransmitted data
> from my thinking, when a retransmit occures, and there is some data buffered ahead
> once the gap is filled data is read from buffer but the sender doesn't know and continues to send in sequence so for the already buffered size packets should be dropped again
> by this I'd expect another triplicate ack to be triggered
> not sure how to deal with this

the triplicate ack basically tells the sender that some data got missed. either move the triplicate ack detection to the receiver and flag the Triplicate Ack or tell back the sender the sequence number of the received packet to filter out acks when the ack is ahead

### PR DESCRIPTION
**Pull request by everHannes**
Fixes #15 
based on #18

### CODE REVIEW

File: utcp.c [hunk 1]
```diff
@@ -1227,11 +1389,13 @@ ssize_t utcp_recv(struct utcp *utcp, const void *data, size_t len) {
 
     size_t datalen = len;
     bool acceptable = false;
+    bool ahead = false;
```
Comment (a-krebs) [inline utcp.c]: This variable is unused.

Comment (everHannes) [reply utcp.c]: true, forgot to remove it

**Comment by a-krebs** [comment]
There's a check in `utcp_recv` that drops packets with unknown flags. That check needs to be updated for `RTR`.

I'm hesitant about adding the `RTR` control flag. It isn't part of any TCP specification I can find. If I understand #15 correctly, the problem is that the sender continues retransmitting data that is already in the receiver's buffer after a buffer gap is filled? Shouldn't the receiver ACK the end-post byte in the receive buffer once a gap is filled, and the sender then effectively stops retransmitting?

**Comment by everHannes** [comment]
> There's a check in utcp_recv that drops packets with unknown flags. That check needs to be updated for RTR.

ah, right

> I'm hesitant about adding the RTR control flag. It isn't part of any TCP specification I can find. If I understand #15 correctly, the problem is that the sender continues retransmitting data that is already in the receiver's buffer after a buffer gap is filled?

Yep it is an issue that came up by the receive buffer in that packets that follow the filled gap potentially raise another triplicate ack since the data already was buffered. I didn't search much for a proposed solution but since only once a packet is received it really is clear if it's buffered already three solutions came to my mind:
1) send back the received sequence number together with the ack so it can be filtered when lower than the ack
2) count for the triplicate ack on the receiver side and flag to explictly request a retransmit for a missing packet
3) or flag when to count a packet ack for a possible retransmit, which I choosed since it was easy to implement and doesn't require any additional header space

Alternatively we could rely on the SACKs to avoid resending packets that already are buffered. However that info might come in delayed ( since the sender skips those that already were sack'ed ) or might get missed completely. In practice this shouldn't be much of a problem though unless the cwnd becomes very large.
But Guus already started work on it so I didn't want to run for my own implementation here, see #16

When you prefer 1) I think that would be the cleanest and a reasonable info what seqno exactly got acked.

> Shouldn't the receiver ACK the end-post byte in the receive buffer once a gap is filled, and the sender then effectively stops retransmitting?

Sure but by then the sender easily sent enough packets to raise another triplicate ack. Not that much of a problem either unless frequently packets are dropped but since that's a small optimization easy to implement why not?
Like you already found there are lots of TCP implementations with slight derivations for optimization. I don't see why utcp needs to follow the RFC documents to the point but rather interpreted it as a source for a well prooven algorithm.

But yeah I'm open to any alternative implementations. It's just that this issue of the filling gap and possible double triplicate ack needs to be addressed somehow.

**Comment by a-krebs** [comment]
> But yeah I'm open to any alternative implementations. 

I couldn't think of one, unfortunately.

**Pull request merged**